File: orquestator/api/session/session.go
package session

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
)

type AgentResponse struct {
	Action string          `json:"action"`
	Data   json.RawMessage `json:"data"`
}

// Starts a new session using a callSid
func Start(callSid string) (*AgentResponse, error) {
	apiURL := os.Getenv("AGENT_API") + "/api/session/start"
	params := url.Values{}
	params.Add("call_sid", callSid)

	fullURL := fmt.Sprintf("%s?%s", apiURL, params.Encode())

	resp, err := http.Post(fullURL, "application/json", nil)
	if err != nil {
		return nil, fmt.Errorf("Error calling LLM endpoint: %v", err)
	}
	defer resp.Body.Close()

	var agentResponse AgentResponse
	if err := json.NewDecoder(resp.Body).Decode(&agentResponse); err != nil {
		return nil, fmt.Errorf("Error parsing JSON: %v", err)
	}

	return &agentResponse, nil
}

// Send message to the session as a User
func Send(callSid string, message string) (*AgentResponse, error) {
	apiURL := os.Getenv("AGENT_API") + "/api/session/send"
	params := url.Values{}
	params.Add("call_sid", callSid)
	params.Add("message", message)

	fullURL := fmt.Sprintf("%s?%s", apiURL, params.Encode())

	resp, err := http.Post(fullURL, "application/json", nil)
	if err != nil {
		return nil, fmt.Errorf("Error calling LLM endpoint: %v", err)
	}
	defer resp.Body.Close()

	var agentResponse AgentResponse
	if err := json.NewDecoder(resp.Body).Decode(&agentResponse); err != nil {
		return nil, fmt.Errorf("Error parsing JSON: %v", err)
	}

	return &agentResponse, nil
}

// Adds context for the model
func Context(callSid string, context string) (*AgentResponse, error) {
	apiURL := os.Getenv("AGENT_API") + "/api/session/context"
	params := url.Values{}
	params.Add("call_sid", callSid)
	params.Add("context", context)

	fullURL := fmt.Sprintf("%s?%s", apiURL, params.Encode())

	resp, err := http.Post(fullURL, "application/json", nil)
	if err != nil {
		return nil, fmt.Errorf("Error calling LLM endpoint: %v", err)
	}
	defer resp.Body.Close()

	var agentResponse AgentResponse
	if err := json.NewDecoder(resp.Body).Decode(&agentResponse); err != nil {
		return nil, fmt.Errorf("Error parsing JSON: %v", err)
	}

	return &agentResponse, nil
}

// Ends and delete a existing session
func End(callSid string) error {
	apiURL := os.Getenv("AGENT_API") + "/api/session/end"
	params := url.Values{}
	params.Add("call_sid", callSid)

	fullURL := fmt.Sprintf("%s?%s", apiURL, params.Encode())

	req, err := http.NewRequest(http.MethodDelete, fullURL, nil)
	if err != nil {
		return err
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	return nil
}

File: orquestator/api/session/session.py
import httpx
import os
from typing import Dict, Any

AGENT_API = os.getenv("AGENT_API")

async def start_session(call_sid: str) -> Dict[str, Any]:
    """Inicia una nueva sesión"""
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{AGENT_API}/api/session/start",
            params={"call_sid": call_sid}
        )
        response.raise_for_status()
        return response.json()

async def send_message(call_sid: str, message: str) -> Dict[str, Any]:
    """Envía un mensaje a la sesión"""
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{AGENT_API}/api/session/send",
            params={"call_sid": call_sid, "message": message}
        )
        response.raise_for_status()
        return response.json()

async def add_context(call_sid: str, context: str) -> Dict[str, Any]:
    """Agrega contexto a la sesión"""
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{AGENT_API}/api/session/context",
            params={"call_sid": call_sid, "context": context}
        )
        response.raise_for_status()
        return response.json()

async def end_session(call_sid: str):
    """Termina y elimina una sesión"""
    async with httpx.AsyncClient() as client:
        response = await client.delete(
            f"{AGENT_API}/api/session/end",
            params={"call_sid": call_sid}
        )
        response.raise_for_status()